registry/handlers/api_test.go

```diff
@@ -520,7 +520,11 @@ func doPushLayer(t *testing.T, ub *v2.URLBuilder, name string, dgst digest.Diges
 
 // pushLayer pushes the layer content returning the url on success.
 func pushLayer(t *testing.T, ub *v2.URLBuilder, name string, dgst digest.Digest, uploadURLBase string, body io.Reader) string {
-	resp, err := doPushLayer(t, ub, name, dgst, uploadURLBase, body)
+	buf := new(bytes.Buffer)
```

Comment:
I see now: body is needed for `doPushLayer` and `digest.FromBytes`. Rather than copy the whole, potentially large buffer, you can use an `io.TeeReader`:

```
dgster = digest.NewDigesterDefault()
resp, err := doPushLayer(t, ub, name, dgst, uploadURLBase,  io.TeeReader(body, dgster))
...
dgst := dgster.Digest()
```

We'll have to add this to the digest package:

```
// Digester calculates the digest of written data. It is functionally
// equivalent to hash.Hash but provides methods for returning the Digest type
// rather than raw bytes.
type Digester struct {
	alg  string
	hash hash.Hash
}

func NewDigester(alg string, h hash.Hash) Digester {
	return Digester{
		alg:  alg,
		hash: h,
	}
}

func NewDigesterDefault() Digester {
	return NewDigester("sha256", sha256.New())
}

// Write data to the digester. These writes cannot fail.
func (d *Digester) Write(p []byte) (n int, err error) {
	return d.hash.Write(p)
}

// Digest returns the current digest for this digester.
func (d *Digester) Digest() Digest {
	return NewDigest(d.alg, d.hash)
}

// Reset the state of the digester.
func (d *Digester) Reset() {
	d.hash.Reset()
}
```